interop, op-supervisor, op-node: Managed Mode

protolambda opened the pull request:
# What

Introduces `ManagedMode` as a major feature branch. Has the following high level features:

- `Controller` module for Supervisor to listen to Node events and deliver commands to Managed Node
- Replacement of the old interop deriver in favor of `Managed Mode` deriver behaviors
- L1 tracking handled by the Supervisor

Note: Standard Mode (the counter to Managed Mode) is still in development and is not available for interop

**Tests**
- CI checks all pass
- E2E tests have been updated
- Unit tests have been updated
- New Unit tests for Controller code

See design doc for more context: https://github.com/ethereum-optimism/design-docs/blob/main/protocol/supervisor-dataflow.md

**Metadata**

Fix https://github.com/ethereum-optimism/optimism/issues/13336

axelKingsley commented on op-node/node/node.go:
```diff
@@ -430,7 +430,7 @@ func (n *OpNode) initL2(ctx context.Context, cfg *Config) error {
 		n.safeDB = safedb.Disabled
 	}
 	n.l2Driver = driver.NewDriver(n.eventSys, n.eventDrain, &cfg.Driver, &cfg.Rollup, n.l2Source, n.l1Source,
-		n.supervisor, n.beacon, n, n, n.log, n.metrics, cfg.ConfigPersistence, n.safeDB, &cfg.Sync, sequencerConductor, altDA)
+		n.beacon, n, n, n.log, n.metrics, cfg.ConfigPersistence, n.safeDB, &cfg.Sync, sequencerConductor, altDA)
```
Although it takes more LoC, could we please linearize the parameters? When we've got so many inputs to the driver  it can be hard to see the order. We've got 6 inputs on the first line and 10 on the next. I see this style in other places in the codebase, so I recognize there is a precedent, but I wanted to offer the suggestion anyway.

axelKingsley replied:
(because I am merging and resolving conflicts on this line anyway, I'll be adopting my suggestion. Feel free to argue your way, happy to hear more about your style preferences)

protolambda replied:
Ideally we get rid of the driver-constructor all the way. There's no point in bundling if we split it into sub-systems nicely later. For now it's fine to linearize.

axelKingsley commented on op-node/rollup/interop/standard/system.go:
```diff
@@ -0,0 +1,69 @@
+	defer s.mu.Unlock()
+
+	switch x := ev.(type) {
+	case rollup.ResetEvent:
+		s.log.Error("todo: interop needs to handle resets", x.Err)
+		// TODO(#13337): on reset: consolidate L2 against supervisor, then do force-reset
+	}
```
I'm not sure why a Standard Mode Subsystem would need to do anything on Reset. It appears to be stateless

axelKingsley reviewed (commented):
Annotating the changes as I see them (mostly for my own benefit)

- The supervisor and server are now officially replaced by an interop SubsSystem!
- Real `Setup` is used and is modified to include the rollup Config, and L1/L2 sources and creates the real Managed/Standard backends
- Managed Mode Implementation
  - Creates an RPC server to accept requests
  - Can Start only once an emitter is attached
  - Only responds to `UnsafeUpdateEvent` Event currently (not sure if more are needed)
  - Most functions are hooked to the RPC API
- Standard Mode Implementation is basically unimplemented atm.

I'm going to merge this with the Managed Mode L1 Traversal: https://github.com/ethereum-optimism/optimism/pull/13392

EDIT: This review was done before we decided to build this as a feature branch here, so many of the annotations are incorrect and the mention about merging it is also not accurate.

tynes commented on op-node/rollup/interop/managed/api.go:
```diff
@@ -0,0 +1,56 @@
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	supervisortypes "github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+)
+
```
Is this API in the spec?

axelKingsley replied:
Managed Nodes (and their API) are not yet in the spec. I would consider Managed mode to be more of an implementation detail of the Supervisor than a spec behavior of the node. If we wanted other clients to support a Managed Mode, we'll need to describe the total set of behaviors that need to change, which is still evolving as we land it.

protolambda committed:
interop: dataflow refactor v2

Co-authored-by: Axel Kingsley <axel.kingsley@gmail.com>

protolambda commented:
Squashed and rebased on develop. Working on getting tests running

joshklop commented on op-node/rollup/interop/config.go:
```diff
@@ -30,55 +33,27 @@ type Config struct {
-	//if (cfg.SupervisorAddr == "") != (cfg.RPCAddr == "") {
-	//	return errors.New("must have either a supervisor RPC endpoint to follow, or interop RPC address to serve from")
-	//}
+	if (cfg.SupervisorAddr == "") != (cfg.RPCAddr == "") {
```
```suggestion
	if (cfg.SupervisorAddr == "") == (cfg.RPCAddr == "") {
```
Based on the error message and other code, it seems this should be `==`.

protolambda committed:
interop: many refactor-fixes, make action test pass

protolambda commented:
Added a commit with a bunch of fixes; use DerivedPair types, fix borked local-safe updates, fix L1-accessor client attachment, etc.

And made the attach-sync-source method on the supervisor return an interface, to await when it sends back up to a given block update to the op-node. The controller-tests are broken again however.

Now the interop action-test passes, but it flakes.

protolambda committed:
op-e2e: fix lint

protolambda committed:
op-service: RPC event-stream util, for subscription and polling flexibility

protolambda committed:
op-node: update interop managed-mode RPC events to use RPC stream util

protolambda committed:
op-supervisor: subscribe or poll for managed-node events

protolambda committed:
op-service: stream fallback util

protolambda committed:
op-node: fix interop managed L1 traversal

protolambda committed:
op-supervisor: enable action tests to pull events

protolambda committed:
op-e2e: fix interop action test

protolambda commented:
Pushed some changes to make the action test not flaky anymore, fix some bugs in L1 traversal in interop-mode, and make polling of events / HTTP connection possible.

axelKingsley committed:
lint

axelKingsley commented:
Fixed lint and rebased. there is a failing test still to be diagnosed, looking into that now.

axelKingsley committed:
fix unit test

axelKingsley committed:
Fixes from E2E tests
- Reset Engine even when in Managed Mode
- Use correct wrapped subscription
- Export wrapped subscription type
- Close Controllers on exit

axelKingsley committed:
Controller tests and fixes

protolambda reviewed (commented):
lgtm

axelKingsley reviewed (approved):
(this PR features commits from both Proto and myself. We have reviewed each others' commits and so I am approving this)
